service/spi/src/main/java/org/kaazing/gateway/service/collections/MemoryTopic.java

```diff
@@ -0,0 +1,133 @@
+/**
+ * Copyright 2007-2016, Kaazing Corporation. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.kaazing.gateway.service.collections;
+
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.hazelcast.core.ITopic;
+import com.hazelcast.core.Message;
+import com.hazelcast.core.MessageListener;
+import com.hazelcast.monitor.LocalTopicStats;
+import com.hazelcast.monitor.impl.LocalTopicStatsImpl;
+
+public class MemoryTopic<E> extends MemoryDistributedObject implements ITopic<E> {
+
+    private static final ExecutorService AR_EXECUTOR =
+        Executors.newSingleThreadExecutor(runnable -> new Thread(runnable, "memory-topics-add/remove"));
+    private static final ExecutorService P_EXECUTOR =
```

Comment:
Using two threads, and especially using ExecutorService and submitting tasks will create garbage, and is using an unbounded queue. I would prefer if you can move to an architecture closer to that of Nuklei and the MQ driver, where each work request (add listener, remove listener and publish) is submitted to a fixed length Agrona queue (instance of ManyToOneConcurrentArrayQueue), and executed by a single worker thread. The registration id returned by addMessageListener would be included in the enqueued setMessageListener command so the method does not need to get an answer from the worker thread. You could have a single worker thread to service all the ITopic instances if you want, though I think it's fine to have a separate thread per topic if it's easier, since we will not be needing to use a large number of topics.